File: UniversalPassportReader/Sources/UniversalPassportReader/NFC/ASN1Parser.cpp
#include "ASN1Parser.hpp"
#include <iostream>

std::vector<TLVNodeCpp> ASN1ParserCpp::parse(const std::vector<uint8_t>& bytes) {
    std::vector<TLVNodeCpp> nodes;
    int index = 0;
    int bytes_count = (int)bytes.size();
    
    while (index < bytes_count) {
        // Skip padding bytes (filler 0x00 or 0xFF)
        while (index < bytes_count && (bytes[index] == 0x00 || bytes[index] == 0xFF)) {
            index++;
        }
        if (index >= bytes_count) break;
        
        // Read Tag
        int tag_start = index;
        uint8_t first_tag_byte = bytes[index];
        index++;
        
        if ((first_tag_byte & 0x1F) == 0x1F) {
            // Multi-byte tag
            while (index < bytes_count) {
                uint8_t next_byte = bytes[index];
                index++;
                if ((next_byte & 0x80) == 0) {
                    break;
                }
            }
        }
        std::vector<uint8_t> tag(bytes.begin() + tag_start, bytes.begin() + index);
        
        // Read Length
        if (index >= bytes_count) break;
        uint8_t first_len_byte = bytes[index];
        index++;
        
        int length = 0;
        if ((first_len_byte & 0x80) == 0) {
            length = first_len_byte;
        } else {
            int num_len_bytes = first_len_byte & 0x7F;
            if (index + num_len_bytes <= bytes_count) {
                for (int i = 0; i < num_len_bytes; i++) {
                    length = (length << 8) | bytes[index];
                    index++;
                }
            } else {
                break;
            }
        }
        
        // Read Value
        if (index + length <= bytes_count) {
            std::vector<uint8_t> value(bytes.begin() + index, bytes.begin() + index + length);
            index += length;
            nodes.push_back({tag, length, value});
        } else {
            std::vector<uint8_t> value(bytes.begin() + index, bytes.end());
            index = bytes_count;
            nodes.push_back({tag, (int)value.size(), value});
            break;
        }
    }
    
    return nodes;
}

bool ASN1ParserCpp::find_node(uint32_t tag, const std::vector<TLVNodeCpp>& nodes, TLVNodeCpp& out_node) {
    for (const auto& node : nodes) {
        if (node.tag_value() == tag) {
            out_node = node;
            return true;
        }
        
        // Check constructed tag (bit 6 is 1)
        if (!node.tag.empty() && (node.tag[0] & 0x20) == 0x20) {
            std::vector<TLVNodeCpp> children = parse(node.value);
            if (find_node(tag, children, out_node)) {
                return true;
            }
        }
    }
    return false;
}

static int find_signature(const std::vector<uint8_t>& signature, const std::vector<uint8_t>& bytes) {
    if (bytes.size() < signature.size()) return -1;
    for (size_t i = 0; i <= bytes.size() - signature.size(); i++) {
        bool found = true;
        for (size_t j = 0; j < signature.size(); j++) {
            if (bytes[i + j] != signature[j]) {
                found = false;
                break;
            }
        }
        if (found) {
            return (int)i;
        }
    }
    return -1;
}

std::vector<uint8_t> ASN1ParserCpp::extract_jpeg_bytes(const std::vector<uint8_t>& bytes) {
    // JPEG start-of-image (FF D8 FF)
    std::vector<uint8_t> jpeg_sig = {0xFF, 0xD8, 0xFF};
    int idx = find_signature(jpeg_sig, bytes);
    if (idx != -1) {
        return std::vector<uint8_t>(bytes.begin() + idx, bytes.end());
    }
    
    // JPEG 2000 start-of-image (00 00 00 0C 6A 50 20 20)
    std::vector<uint8_t> jp2_sig = {0x00, 0x00, 0x00, 0x0C, 0x6A, 0x50, 0x20, 0x20};
    idx = find_signature(jp2_sig, bytes);
    if (idx != -1) {
        return std::vector<uint8_t>(bytes.begin() + idx, bytes.end());
    }
    
    return std::vector<uint8_t>();
}

File: UniversalPassportReader/Sources/UniversalPassportReader/NFC/ASN1Parser.swift
import Foundation

public struct TLVNode {
    public let tag: [UInt8]
    public let length: Int
    public let value: [UInt8]
    
    public var tagValue: UInt32 {
        var val: UInt32 = 0
        for byte in tag {
            val = (val << 8) | UInt32(byte)
        }
        return val
    }
}

public class ASN1Parser {
    
    public static func parse(bytes: [UInt8]) -> [TLVNode] {
        var nodes: [TLVNode] = []
        var index = 0
        
        while index < bytes.count {
            // Skip padding bytes (filler 0x00 or 0xFF)
            while index < bytes.count && (bytes[index] == 0x00 || bytes[index] == 0xFF) {
                index += 1
            }
            if index >= bytes.count { break }
            
            // Read Tag
            let tagStartIndex = index
            let firstTagByte = bytes[index]
            index += 1
            
            // If bits 1-5 are 1, tag is multi-byte
            if (firstTagByte & 0x1F) == 0x1F {
                while index < bytes.count {
                    let nextByte = bytes[index]
                    index += 1
                    // If MSB is 0, this is the last byte of the tag
                    if (nextByte & 0x80) == 0 {
                        break
                    }
                }
            }
            let tag = Array(bytes[tagStartIndex..<index])
            
            // Read Length
            if index >= bytes.count { break }
            let firstLenByte = bytes[index]
            index += 1
            
            var length = 0
            if (firstLenByte & 0x80) == 0 {
                // Short form length (0-127)
                length = Int(firstLenByte)
            } else {
                // Long form length
                let numLenBytes = Int(firstLenByte & 0x7F)
                if index + numLenBytes <= bytes.count {
                    for _ in 0..<numLenBytes {
                        length = (length << 8) | Int(bytes[index])
                        index += 1
                    }
                } else {
                    break
                }
            }
            
            // Read Value
            if index + length <= bytes.count {
                let value = Array(bytes[index..<(index + length)])
                index += length
                nodes.append(TLVNode(tag: tag, length: length, value: value))
            } else {
                let value = Array(bytes[index..<bytes.count])
                index = bytes.count
                nodes.append(TLVNode(tag: tag, length: value.count, value: value))
                break
            }
        }
        
        return nodes
    }
    
    public static func findNode(tag: UInt32, in nodes: [TLVNode]) -> TLVNode? {
        for node in nodes {
            if node.tagValue == tag {
                return node
            }
            // Check if tag indicates a constructed TLV (bit 6 is 1)
            if node.tag.count > 0 && (node.tag[0] & 0x20) == 0x20 {
                let children = parse(bytes: node.value)
                if let found = findNode(tag: tag, in: children) {
                    return found
                }
            }
        }
        return nil
    }
    
    public static func extractJPEGBytes(from bytes: [UInt8]) -> [UInt8]? {
        // Look for JPEG start-of-image signature (FF D8 FF)
        let jpegSignature: [UInt8] = [0xFF, 0xD8, 0xFF]
        if let idx = findSignature(jpegSignature, in: bytes) {
            return Array(bytes[idx..<bytes.count])
        }
        
        // Look for JPEG 2000 start-of-image signature (00 00 00 0C 6A 50 20 20)
        let jp2Signature: [UInt8] = [0x00, 0x00, 0x00, 0x0C, 0x6A, 0x50, 0x20, 0x20]
        if let idx = findSignature(jp2Signature, in: bytes) {
            return Array(bytes[idx..<bytes.count])
        }
        
        return nil
    }
    
    private static func findSignature(_ signature: [UInt8], in bytes: [UInt8]) -> Int? {
        guard bytes.count >= signature.count else { return nil }
        for i in 0...(bytes.count - signature.count) {
            var found = true
            for j in 0..<signature.count {
                if bytes[i + j] != signature[j] {
                    found = false
                    break
                }
            }
            if found {
                return i
            }
        }
        return nil
    }
}
